### PR TITLE
Johnkoh msft azuredatatransfer microsoft.azure data transfer 2023 10 11 preview

### PR DESCRIPTION
**Pull request by johnkoh-msft**
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [ ] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [x] Other, please clarify:
    - Our API specs have previously been merged into RPSaaSMaster in the PR repository. We are looking to move this new API version to the public repo. To see the changes between this new API version and the last API version that was approved for RPSaaSMaster, please view this closed PR in the PR repo: https://github.com/Azure/azure-rest-api-specs-pr/pull/15345

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by zizw123** [comment]
Please take a look at the lint error https://github.com/Azure/azure-rest-api-specs/pull/26576/checks?check_run_id=18458171799

**Comment by zizw123** [comment]
> Hi @johnkoh-msft! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

Please also take a look at this

**Comment by johnkoh-msft** [comment]
> > Hi @johnkoh-msft! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).
> 
> Please also take a look at this

Added base commit to show changes between existing version in the private repo and this new version in the public repo.

https://github.com/Azure/azure-rest-api-specs/pull/26576/files/9bba3f67d15e91314665be5f924e4cf003330714..dbf8ee3ff6421fd7c234ebb1d8abbb19749a0a5d

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:37]: >action [](http://example.com/codeflow?start=44&length=6)

You have an action called... action? And not something more specific?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1422]: >  "description": "An action request", [](http://example.com/codeflow?start=4&length=37)

and what is that?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1443]: and what would be in this?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1432]: and what would be in this?

**Comment by TimLovellSmith** [comment]
> 

why?

---
Refers to: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/examples/postAction.json:21 in 262f3eb. [](commit_id = 262f3ebc2f84584e0ecf0eb55be0d0d75521a503, deletion_comment = False)

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1446]: >selectedResource [](http://example.com/codeflow?start=5&length=16)

This looks a lot like TrackedResource in common-types but not quite the same. Why the differences?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1506]: >description [](http://example.com/codeflow?start=11&length=11)

"format": "arm-id"?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1748]: >description [](http://example.com/codeflow?start=11&length=11)

"format": "arm-id"?

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:433]: >resourceBody [](http://example.com/codeflow?start=37&length=12)

why is it called a resourceBody and not e.g. a flowBody?

**Review by TimLovellSmith — CHANGES_REQUESTED** [review]
removing wait

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1446]: https://learn.microsoft.com/en-us/azure/azure-resource-manager/managed-applications/microsoft-solutions-resourceselector

This maps to the output of the ResourceSelector UI element which is not the same as the TrackedResource.

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1506]: The resource is not guaranteed to be in the same tenant, so I believe we cannot use "format": "arm-id" with `x-ms-arm-id-details`.

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1443]: The payload for the request/response would need to be dynamic to support the open nature of the post action.

Comment (visingla-ms) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1443]: This an anti-pattern in ARM APIs, as customers won't know what to provide in request payload and there is no client side (SDK side) validation that they can use to build their payload. That's why using additional properties is not allowed. Please sign up for ARM API Modeling Office Hours to discuss your use case: [https://armwiki.azurewebsites.net/introduction/office_hours.html](https://armwiki.azurewebsites.net/introduction/office_hours.html).

Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1441]: The additional properties for the provider-level action have been removed.

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (ms-zhenhua) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1422]: please confirm this comment, and provide more details in description for this property if possible.

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json]: Renamed.

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json]: Updated the `description` to include more information about the request.

**Review by ms-zhenhua — APPROVED** [review]

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1441]: Are you somehow encoding a HTTP response in a string? That seems like a potential worry.

Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1445]: It is a message string describing the result. I have added a longer description and added a `code` property to the result as well to make the purpose of the message clearer.

**Comment by TimLovellSmith** [comment]
Is this a realistic example?
What are the semantics?

>{
  "parameters": {
    "api-version": "2023-10-11-preview",
    "parameters": {
      "name": "actionName",
      "payload": "my requested action"
    }
  },
  "responses": {
    "200": {
      "body": {
        "result": "action completed"
      }
    }
  }
}

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1445]: Can you please further explain how anyone using this API knows the valid inputs and outputs to use for 'perform request'?

Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:1499]: We have updated our API specs to separate out the general action endpoint into more specific provider-level actions.

**Review by ms-henglu — APPROVED** [review]

**Review by ms-henglu — APPROVED** [review]

**Review by ms-henglu — APPROVED** [review]

**Review by zizw123 — APPROVED** [review]

**Comment by zizw123** [comment]
@TimLovellSmith it seems this PR is pending you approval because of the changes requested check, even removing you from the reviewers won't help, please take a look again when you have a chance to unblock this. Thanks!

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (TimLovellSmith) [inline specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:77]: >post [](http://example.com/codeflow?start=7&length=4)

Why do you want to use POST verb for this and not GET? Is there some sort of sensitivity in the data returned?

Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:77]: We are creating a [provider-level action](https://armwiki.azurewebsites.net/rpaas/contract/resourcetypecategory.html#4-providerlevelaction) through RPaaS (an endpoint at the provider level and not at a resource level) which needs to be a POST action.

Comment (TimLovellSmith) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:77]: @johnkoh-msft Is making it a provider-level proxy (singleton?) resource type, instead of a provider-level action, possible? If thats possible wouldn't it be preferable, so that it can be run with read-only permissions?

**Comment by mentat9** [comment]
@johnkoh-msft - Please address open comments before removing the ARMChangesRequested label. This one is still pending: https://github.com/Azure/azure-rest-api-specs/pull/26576#discussion_r1442088218.

File: specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json
Comment (johnkoh-msft) [reply specification/azuredatatransfer/resource-manager/Microsoft.AzureDataTransfer/preview/2023-10-11-preview/azuredatatransfer.json:77]: > @johnkoh-msft Is making it a provider-level proxy (singleton?) resource type, instead of a provider-level action, possible? If thats possible wouldn't it be preferable, so that it can be run with read-only permissions?

This is a good point to consider; however, we believe that the call to "list approved schemas" is more of an "Action" than a "Read" since there are some calculations done on the backend to generate this list. An GET on a [proxy resource](https://armwiki.azurewebsites.net/rpaas/pureproxy.html) called "approvedSchemas" could be possible, but conceptually with our RP, "approvedSchemas" is not associated with a resource definition. We believe a provider-level action aligns more with the action being performed.

**Review by TimLovellSmith — APPROVED** [review]

**Pull request merged**
